app/build_images.go

```diff
@@ -0,0 +1,225 @@
+package app
+
+import (
+	"archive/tar"
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"code.cloudfoundry.org/fissile/builder"
+	"code.cloudfoundry.org/fissile/docker"
+	"code.cloudfoundry.org/fissile/model"
+	"github.com/SUSE/stampy"
+	"github.com/fatih/color"
+)
+
+// BuildImagesOptions contains all option values for the `fissile build images` command.
+type BuildImagesOptions struct {
+	Force                    bool
+	Labels                   map[string]string
+	NoBuild                  bool
+	OutputDirectory          string
+	PatchPropertiesDirective string
+	Roles                    []string
+	Stemcell                 string
+	StemcellID               string
+	TagExtra                 string
+}
+
+// BuildImages builds all role images using releases
+func (f *Fissile) BuildImages(opt BuildImagesOptions) error {
+	err := f.LoadManifest()
+	if err != nil {
+		return err
+	}
+	if f.Manifest == nil || len(f.Manifest.LoadedReleases) == 0 {
+		return fmt.Errorf("Releases not loaded")
+	}
+	if errs := f.Validate(); len(errs) != 0 {
+		return fmt.Errorf(errs.Errors())
+	}
+
+	if opt.OutputDirectory != "" {
+		err := os.MkdirAll(opt.OutputDirectory, 0755)
+		if err != nil {
+			if os.IsExist(err) {
+				return fmt.Errorf("Output directory %s exists and is not a directory", opt.OutputDirectory)
+			}
+			if err != nil {
+				return fmt.Errorf("Error creating directory %s: %s", opt.OutputDirectory, err)
+			}
+		}
+	}
+
+	if f.Options.Metrics != "" {
+		stampy.Stamp(f.Options.Metrics, "fissile", "create-images", "start")
+		defer stampy.Stamp(f.Options.Metrics, "fissile", "create-images", "done")
+	}
+
+	if opt.StemcellID == "" {
+		imageManager, err := docker.NewImageManager()
+		if err != nil {
+			return err
+		}
+
+		stemcellImage, err := imageManager.FindImage(opt.Stemcell)
+		if err != nil {
+			if _, ok := err.(docker.ErrImageNotFound); ok {
+				return fmt.Errorf("Stemcell %s", err.Error())
+			}
+			return err
+		}
+
+		opt.StemcellID = stemcellImage.ID
+	}
+
+	packagesImageBuilder := &builder.PackagesImageBuilder{
+		RepositoryPrefix:     f.Options.RepositoryPrefix,
+		StemcellImageName:    opt.Stemcell,
+		StemcellImageID:      opt.StemcellID,
+		CompiledPackagesPath: f.StemcellCompilationDir(opt.Stemcell),
+		FissileVersion:       f.Version,
+	}
+
+	instanceGroups, err := f.Manifest.SelectInstanceGroups(opt.Roles)
+	if err != nil {
+		return err
+	}
+
+	if opt.OutputDirectory == "" {
+		err = f.buildPackagesImage(opt, instanceGroups, packagesImageBuilder)
+	} else {
+		err = f.buildPackagesTarball(opt, instanceGroups, packagesImageBuilder)
+	}
+	if err != nil {
+		return err
+	}
+
+	imageName, err := packagesImageBuilder.GetImageName(f.Manifest, instanceGroups, f)
+	if err != nil {
+		return err
+	}
+
+	roleImageBuilder := &builder.RoleImageBuilder{
```

Comment:
I actually got rid of the `NewRoleImageBuilder` in this release. I dislike having function with 10 parameters or even more, given that Go doesn't support named parameters. Creating a struct makes the mapping of values to fields much more obvious, so you don't end up passing e.g. your booleans in the wrong order.

There was actually one field that *did* get the wrong value, but it didn't matter because the correct value was passed in later to a `Build` method, and the value from `NewRoleImageBuilder` was never used.

Anyways, I prefer this style, which is more self-documenting than using fake constructors.

Reply:
I agree with the self-documenting. For longer initializers, I've seen and personally have used custom defined types to make it easier to track all the parameters and the propagation. I believe it is still worth it, but up to you!

Reply:
I don't understand what you are saying here, isn't `builder.RoleImageBuilder` a custom defined type? Can you point me to an example?

Looking at the previous version of the constructor, I don't see any advantage to using it over directly constructing the struct at the callers side: https://github.com/cloudfoundry-incubator/fissile/blob/a19161d/builder/role_image.go#L57-L76

How is it "worth it"?